subsys/bluetooth/host/keys.c

```diff
@@ -279,6 +279,24 @@ void bt_keys_add_type(struct bt_keys *keys, int type)
 	keys->keys |= type;
 }
 
+void bt_keys_clear_other_keys_with_identity(uint8_t id, const bt_addr_le_t *id_addr,
+		struct bt_keys *keep_these_keys)
```

Comment:
Hmm, not a huge fan of this function. 

The use case is that we've added multiple keys for the same identity and remote peer address, and we want to clear all but a specific key? 

When do we want to do this? Is it possible to determine that this needs to be done, but adding the `keep_these_keys` in the first place? i.e. can we know this needs to be done before `keep_these_keys` is added in the first place, and clear the other keys at that point? 

Can a remote address with this identity even "legally" have multiple keys? Shouldn't we just only keep 1 `bt_keys` for the { identity, id_addr } pair in the first place?

Reply:
The issue is that the LTK is stored at a time when the IRK has not yet been exchanged, which creates two entries in the key_pool array. One from the previous bonding using the previous bonding's identity address as key, the second with the new bonding's LTK using the random address since there is no IRK yet.

bt_keys_clear_other_keys_with_identity removes older entries with the new bonding's identity address as key, as there must only be one entry with the same identity as key in the key_pool, to prevent bt_smp_request_ltk from finding the previous bonding's LTK.

So we have to somehow prevent multiple entries with the same identity address as key in key_pool.

Another way could be to perhaps temporarily store the LTK in the bt_conn struct until the IRK and the identity address is known?

Calling bt_keys_clear_other_keys_with_identity from smp_ident_addr_info just seemed simpler. But it is an odd function.

I just don't know yet how else to keep key_pool from having two entries with the same identity address in it. Any ideas?